### PR TITLE
Fixed embeddings

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request adds support fixed word embeddings (that is, embeddings that will not be changed as the parameter is changed) on the source or target side. In order to make this go more smoothly, I also refactored Embeddings -- the constructor now takes the parameters it needs explicitly, instead of just taking the relevant dictionaries and opt. If opt were passed, the model would somehow need to decide whether the `-fix_word_vecs_enc` or `-fix_word_vecs_dec` is relevant. This refactoring also allows loading of pretrained embeddings to be done when the object is instantiated, instead of later on in train.py. Future work could also be done to allow pretrained and/or fixed feature vectors.

### CODE REVIEW

**Comment by srush** [comment]
Hmm. So I like this change. But can I request that you go a step further while you are doing it?

- Can you pull `Embeddings` out into its own file in modules. Just call it `Embeddings.py`.
- While you do this, can you split `Embeddings` and `FeatureEmbeddings` into separate classes.
- Instead of passing in `gpus` as an argument, we should maybe have self.pe that be a `parameter` so it gets autoconverted. 
- I think we should pass in `pre_word_vecs` and `fix_word_vecs` to a separate "load_word_vectors" functions.
- Generally I think we should try to keep constructor arguments for key classes to be somewhat shorter.  and also be passed in by keywords or by a dictionary. Having lists of arguments greater than 4 or 5 is messy. 

Let me know if this makes sense.

**Comment by srush** [comment]
(Actually can you maybe hold off on this for a day or to, we are trying to land `torchtext`, I'll let you know when it is merged)

**Comment by bpopeters** [comment]
I have a pretty busy next few days, so we can leave this for then. 

About all the parameters for the constructor, I agree that it looks sort of messy. Using kwargs or similar isn't a bad idea. I'll give it a look.

What is the justification is for splitting the `Embeddings` class in two?  Is there a performance problem with how it is now? The feature-free embedding behavior is just a special case of the general embedding behavior. There's no need to implement this behavior twice in two different places. Additionally, I would either need to reduplicate the code for loading/fixing embeddings between the two classes, or I would need to introduce some sort of class hierarchy. That seems like a needless complication. Finally, wouldn't it make the code more unwieldy elsewhere to have two different classes for handling embeddings? One would need to add some behavior in the encoder/decoder just to decide which of the two classes to call.

**Comment by srush** [comment]
Yeah, understandable. Maybe I'll take a pass at it. 

I think it is an interesting question whether we want to split things up into classes. I initially agreed with you that we shouldn't, but I am worrying that the code is becoming a bit hard to read for people who are not using all the features. For instance here is the embedding class with no features: 

```
class Embeddings(nn.Module):                                                                                                                                                                                      
    def __init__(self, dim, vocab):                                                                                                                                                                          
        super(Embeddings, self).__init__()                                                                                                                                                                        
        self.word_lut = \                                                                                                                                                                                         
                        nn.Embedding(len(vocab), dim,                                                                                                                                                             
                                     padding_idx=dicts.stoi[onmt.IO.PAD_WORD])                                                                                                                                    
                                                                                                                                                                                                                  
    def load_pretrained_vectors(self, emb_file):                                                                                                                                                                  
        if emb_file is not None:                                                                                                                                                                                  
            pretrained = torch.load(emb_file)                                                                                                                                                                     
            self.word_lut.weight.data.copy_(pretrained)                                                                                                                                                           
                                                                                                                                                                                                                  
    def forward(self, src_input):                                                                                                                                                                                 
        """                                                                                                                                                                                                       
        Return the embeddings for words, and features if there are any.                                                                                                                                           
        """                                                                                                                                                                                                       
        return self.word_lut(src_input.squeeze(2))                                                                                                                                                                 
 ``

I think we could reasonably have that be the base class and inherit a FeatureEmbeddings class in modules without duplicating too much code. We can then have a switch statement when constructing the models to decide which one to use. 

The benefit in my mind is that new users will be able to understand the simple code very quickly.

**Comment by bpopeters** [comment]
That's a reasonable point. Later this week, I'll take another look at it and see how understandable I can make the interface.

**Comment by guillaumekln** [comment]
I think the pretrained embeddings were loaded after instantiation because the parameters are uniformly initialized also after the model instantiation.

See https://github.com/OpenNMT/OpenNMT-py/blob/master/train.py#L273-L274

This was an issue at some point in the Lua version.

**Comment by bpopeters** [comment]
I did not split `Embeddings` into multiple classes, as suggested. However, I refactored it with `nn.Sequential` in a way that I think makes the code easier to follow. 

Noting what Guillaume said, I also moved the pretrained embedding loading and fixing to after the param init. I haven't been able to _rigorously_ test the pretrained vector fixing yet; however, the differences between it and the previous pretrained vector code are minimal.

I needed to make some changes to the `build_embeddings` code because features cannot use the same padding index as the word vocabulary.

**Comment by bpopeters** [comment]
I'll figure out what's going on with the failing tests in the morning.

**Comment by srush** [comment]
> I needed to make some changes to the build_embeddings code because features cannot use the same padding index as the word vocabulary.

Oh yes. This is annoying. Maybe we can enforce this in preprocessing. Passing around multiple paddings is ugly. 

> I'll figure out what's going on with the failing tests in the morning.

The test is failing because one of the weight names changed. Unfortunately this breaks old models. We can either retrain the test models or write our own `__getstates__`

File: onmt/modules/Embeddings.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+from onmt.modules import BottleLinear
+from onmt.modules import aeq
+
+
+class Elementwise(nn.ModuleList):
```
Comment (srush) [inline onmt/modules/Embeddings.py]: Could put this in modules/utils if you think it is useful in other locations.

File: onmt/modules/Embeddings.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            return outputs[0]
+        elif self.merge == 'concat' or self.merge == 'mlp':
+            return torch.cat(outputs, 2)
+        elif self.merge == 'sum':
+            return sum(outputs)
+        else:
+            return outputs
+
+
+class PositionalEncoding(nn.Module):
```
Comment (srush) [inline onmt/modules/Embeddings.py]: Nice, I like that this is a separate class. It seems like people are also using LUT based positions, so this makes it easy to add that. 
Q: will `self.pe` work here with cuda too?

**Comment by srush** [comment]
Okay, this looks good to me. I just have one request. 

For a user who doesn't care about features, it is still a bit hard to find what is "required". Can you explicitly fence off all the code at the top and the bottom of `__init__` with if statements indicating that this is only code with source features. i.e. instead of 

`pad_indices = [padding_idx] + feat_pads` =>

`pad_indices = [padding_idx]`

and then in this one if statement add `feat_pads` (and similarly for other parts). 

My argument is this: I know your code is cleaner and shorter. But the metric I am targeting is not LOC, it is expected LOC a new user needs to find the answer. Fencing off the feature parts tells them they can skip it entirely when reading for the common case, but as it is now it is hard to know whether all this is necessary to read.

File: onmt/modules/Embeddings.py (hunk 1 above)
Comment (bpopeters) [reply onmt/modules/Embeddings.py]: I considered this. I can't think of any other cases where it's useful yet, but it wouldn't be surprising to me if there were some. I'll devote some thought to this and other places in the code where containers could be useful soon.

File: onmt/modules/Embeddings.py (hunk 2 above)
Comment (bpopeters) [reply onmt/modules/Embeddings.py]: Hmm, I thought it would because the `PositionalEncoding` instance is ultimately one of the submodules of the NMTModel. But now I'm not convinced. I think it may need to be registered as a buffer.

**Comment by bpopeters** [comment]
I tried to make the Embeddings code more user-friendly. I tried to do so in the spirit of your suggestion, but I departed from it on a couple of particulars. I did not add any if statements. Instead I visually separated the parts for gathering the word and feature parameters, and I wrote comments to explain the purpose of each block.

**Comment by bpopeters** [comment]
I'm not completely certain that treating PositionalEncoding.pe as a buffer is the right thing to do; maybe somebody who knows a lot about the pytorch particulars could shed some light on that.

**Comment by srush** [comment]
Okay, looks good to me. I think buffer is correct. 

As written this will break old models. If would consider keeping `emb_luts` attached to Embeddings as opposed to `make_embeddings` and getting rid of the Sequential, I think this will pass.

**Comment by srush** [comment]
Other option is to retrain the test_models. One is described in "data/README.md" the other is the model you sent me.

**Comment by bpopeters** [comment]
Personally, I like Sequential because it means we can keep all the business about the module's parameters inside the constructor. If we went back to the old way, then there would need to be for example two checks about whether there is position encoding: one in the constructor, when the submodule is created; and the other in `forward`, when it needs to decide whether to apply position encoding to the embeddings. I think Sequential makes the code clearer this way, although I can change it back if necessary.

**Comment by srush** [comment]
Yeah. I don't have a strong preference, but I would like the test to work.

The argument against sequential is that the design philosophy of pytorch is to separate parameters from control flow. That way if control flow changes the same contructor can be used, or even repeatedly used. While I agree you save an if statement, you then force make_embeddings to own the parameters.

Maybe there is a a way to get both these advantages? Keep parameters on the parent but prespecify the pipeline?

**Comment by bpopeters** [comment]
I understand the argument about design philosophy and agree with it in general. However, I don't think it's an argument against `Sequential`, per se. `Sequential` can be really useful if we have a simple sequence of operations to perform together, like a feed-forward network/layer (see the mlp in embeddings, or the generator). 

I think that's what we have with `make_embedding`. It's a generalization of table look-up. I wouldn't try to use `Sequential` for, say, input feed in the docoder, but I think it makes sense here.

**Comment by srush** [comment]
Okay sounds good. That decision means we have to retrain all the models and samples though. I'll make a script to do that. Can you send the training data for your morphology model?

**Comment by bpopeters** [comment]
[train.zip](https://github.com/OpenNMT/OpenNMT-py/files/1272308/train.zip)

Here's the training data for the Serbo-Croat grapheme-to-phoneme model that I sent before.

**Comment by bpopeters** [comment]
Is there anything more I should do to get this up and running? It would be nice to get it done because #197 broke feature embeddings.

**Comment by srush** [comment]
You're welcome to fix the tests yourself. Otherwise I'll get to it today or
tomorrow.

**Comment by srush** [comment]
Okay pushed. Can you run `bash test/rebuild_test_models.sh` and then `bash test/pull_request_chk.sh`

**Comment by srush** [comment]
Would love to get this merged soon. Maybe I'll fork and fix the tests?

**Comment by bpopeters** [comment]
Sorry for not following up with this quicker. I'll try to get it down now.

**Comment by bpopeters** [comment]
Okay, so I merged the changes in main, but it looks like I won't be able to rebuild the models myself. I don't have a gpu at the moment, so it'll take a while to train the models. For the future, maybe we could shrink the test models a bit? My Serbo-Croat g2p model should still be a good indicator with half of the training data.

**Comment by srush** [comment]
yes can do. (If you need a gpu we can hook you up with one for opennmt work. )
